### PR TITLE
[SPARK-18815] [SQL] Fix NPE when collecting column stats for string/binary column having only null values

### PR DESCRIPTION
**Pull request by wzhfy**
## What changes were proposed in this pull request?

During column stats collection, average and max length will be null if a column of string/binary type has only null values. To fix this, I use default size when avg/max length is null.

## How was this patch tested?

Add a test for handling null columns

### CODE REVIEW

**Comment by wzhfy** [comment]
cc @cloud-fan @rxin

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 1]
```diff
@@ -133,6 +133,79 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
     }
   }
 
+  test("column stats round trip serialization") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: why did you move the test cases?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala [hunk 2]
```diff
@@ -213,7 +214,8 @@ object ColumnStat extends Logging {
         val nullLit = Literal(null, col.dataType)
         struct(
           ndv, nullLit, nullLit, numNulls,
-          Ceil(Average(Length(col))), Cast(Max(Length(col)), LongType))
+          Coalesce(Seq(Ceil(Average(Length(col))), defaultSize)),
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala]: add a line of comment explaining we set the value to defaultSize if all the values are null (of if there is no value).

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 3]
```diff
@@ -133,6 +133,79 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
...
+
+  test("analyze column command - result verification") {
+    val tableName = "column_stats_test2"
+    // (data.head.productArity - 1) because the last column does not support stats collection.
+    assert(stats.size == data.head.productArity - 1)
+    val df = data.toDF(stats.keys.toSeq :+ "carray" : _*)
+    checkColStats(df, tableName, stats)
+  }
+
+  private def checkColStats(
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: Need to add a line of comment documenting this function and change the signature to 

```
/**
  * Compute stats for the given DataFrame and compare it with colStats.
  */
private def checkStats(df: DataFrame, colStats: LinkedHashMap[String, ColumnStat]): Unit = {
}
```

Just generate a random tableName.

Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: Also don't put this in the middle of the function. Move it into the abstract class StatisticsCollectionTestBase

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 4]
```diff
@@ -133,6 +133,79 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
...
+      }
+    }
+  }
+
+  test("column stats collection for null columns") {
+    def nullColumnStat(dataType: DataType): ColumnStat = {
+      ColumnStat(0, None, None, 1, dataType.defaultSize.toLong, dataType.defaultSize.toLong)
+    }
+
+    val tableName = "column_stats_test3"
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: You can simplify this a lot by doing it programmatically, e.g.

```

val dataTypes = Seq(BooleanType, ByteType, ...)
val df = sql("select " + dataTypes.map(tpe => s"cast(null as ${tpe.sql})").mkString(", "))

val expectedStats = dataTypes.map { tpe =>
  ColumnStat(0, None, None, 1, dataType.defaultSize.toLong, dataType.defaultSize.toLong)
}

...
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala (hunk 2 above)
Comment (wzhfy) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala]: ok

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala (hunk 3 above)
Comment (wzhfy) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: ok, I'll fix it

Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: +1

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala (hunk 4 above)
Comment (wzhfy) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: That's much better, thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 5]
```diff
@@ -180,35 +214,28 @@ abstract class StatisticsCollectionTestBase extends QueryTest with SQLTestUtils
...
-    val df = data.toDF(stats.keys.toSeq :+ "carray" : _*)
+  private val randomName = new Random(31)
 
+  /**
+   * Compute column stats for the given DataFrame and compare it with colStats.
+   */
+  def checkColStats(
+      df: DataFrame,
+      colStats: mutable.LinkedHashMap[String, ColumnStat]): Unit = {
+    val tableName = "column_stats_test_" + randomName.nextInt(10)
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: the chance of collision here is not low. just use a broader key space nextInt

Comment (wzhfy) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: ok, done.

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 6]
```diff
@@ -180,35 +214,28 @@ abstract class StatisticsCollectionTestBase extends QueryTest with SQLTestUtils
...
-    val df = data.toDF(stats.keys.toSeq :+ "carray" : _*)
+  private val randomName = new Random(31)
 
+  /**
+   * Compute column stats for the given DataFrame and compare it with colStats.
+   */
+  def checkColStats(
+      df: DataFrame,
+      colStats: mutable.LinkedHashMap[String, ColumnStat]): Unit = {
+    val tableName = "column_stats_test_" + randomName.nextInt(1000)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: why not just use a fixed name like `stats_test_tbl`? It will be used in `withTable`, which means we will drop this table every time.

Comment (wzhfy) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala]: It's in the parent class, if another subclass suite runs in a different thread and also calls this function, there's a chance of collision. Currently it's only called in one suite, but maybe some other suites will use it in the future, so it's safer to use a random name.

**Comment by wzhfy** [comment]
retest this please

**Comment by rxin** [comment]
Merging in master/branch-2.1. Thanks.
